src/Futhark/Pass/CoalesceAccess.hs

```diff
@@ -0,0 +1,303 @@
+        (Loop bindings loop body) -> replaceArrayNamesInLoop arr_map bindings loop body
+        (Op op) -> replaceArrayNamesInOp arr_map op
+        (BasicOp op) -> replaceArrayName arr_map patternName op
+        _ -> e
+  Let pat aux e'
+
+replaceArrayNamesInMatch :: ArrayNameReplacements -> [SubExp] -> [Case (Body GPU)] -> Body GPU -> MatchDec (BranchType GPU) -> Exp GPU
+replaceArrayNamesInMatch arr_map s cases body m = do
+  let bodies = map caseBody cases
+  let body' = replaceArrayNamesInBody arr_map body
+  let cases' =
+        foldl
+          ( \cases_acc case' -> do
+              let body'' = caseBody case'
+              let new_body = replaceArrayNamesInBody arr_map body''
+              let new_case = Case (casePat case') new_body
+              cases_acc ++ [new_case]
+          )
+          mempty
+          cases
+  Match s cases' body' m
+
+replaceArrayNamesInLoop :: ArrayNameReplacements -> [(FParam GPU, SubExp)] -> LoopForm -> Body GPU -> Exp GPU
+replaceArrayNamesInLoop arr_map bindings loop body = Loop bindings loop (replaceArrayNamesInBody arr_map body)
+
+replaceArrayNamesInOp :: ArrayNameReplacements -> Op GPU -> Exp GPU
+replaceArrayNamesInOp arr_map op = do
+  let op' = case op of
+        (SegOp segop) -> replaceArrayNamesInSegOp arr_map segop
+        (GPUBody types body) -> replaceArrayNamesInGPUBody arr_map types body
+        _ -> op
+  Op op'
+
+replaceArrayNamesInSegOp :: ArrayNameReplacements -> SegOp SegLevel GPU -> Op GPU
+replaceArrayNamesInSegOp arr_map op = do
+  let body = segBody op
+  let body' = replaceArrayNamesInKernelBody arr_map body
+  let op' = case op of
+        (SegMap lvl segSpace types body) -> SegMap lvl segSpace types body'
+        (SegRed lvl segSpace o types body) -> SegRed lvl segSpace o types body'
+        (SegScan lvl segSpace o types body) -> SegScan lvl segSpace o types body'
+        (SegHist lvl segSpace o types body) -> SegHist lvl segSpace o types body'
+  SegOp op'
+
+replaceArrayNamesInGPUBody :: ArrayNameReplacements -> [Type] -> Body GPU -> Op GPU
+replaceArrayNamesInGPUBody arr_map types body = GPUBody types (replaceArrayNamesInBody arr_map body)
+
+replaceArrayName :: ArrayNameReplacements -> VName -> BasicOp -> Exp GPU
+replaceArrayName arr_map idx_name op = do
+  -- Look up the new array name
+  case M.lookup idx_name arr_map of
+    Nothing -> BasicOp op
+    Just arr_name' -> do
+      -- Replace the array name in the BasicOp
+      let op' = case op of
+            (Index _name slice) -> Index arr_name' slice
+            (Update safety _name slice subExp) -> Update safety arr_name' slice subExp
+      BasicOp op'
+
+-- TODO:
+-- 1. Find all permutations in PermutationTable for this SegOp
+-- 2. Insert permutations in the AST
+
+-- | Given an ordering function for `DimIdxPat`, and an IndexTable, return
+-- a PermutationTable.
+permutationTableFromIndexTable :: (DimIdxPat rep -> DimIdxPat rep -> Ordering) -> IndexTable rep -> PermutationTable
```

Comment:
I'm not sure it makes sense to express the desired permutation through a total order on DimIdxPat.  A function `[DimIdxPat rep] -> [DimIdxPat rep]` would be superior.

Reply:
Agreed, but we can't figure out a way to retrieve the permutation of the array, if we do it like this.

Reply:
[isPermutationOf](https://hackage.haskell.org/package/futhark-0.25.5/docs/Futhark-IR-Prop-Rearrange.html#v:isPermutationOf)

Reply:
Alternatively, make this a function that returns a permutation (rather than a new list).

Reply:
tldr: I found a better way:)

isPermutationOf is insufficient.
Given
```
segmap (i < n - 1)
  xs [i][i+1]
```
would result in one index expression on two dimensions, with identical loop/kernelthread dependencies (because we  ignore constants and variables that are not dependent on iterators).
They're indistinguishable to `isPermutationOf`, as well as to the sorting function, therefore `isPermutationOf` is likely going to yield the incorrect permutation in most cases.

IMO it would be better to add an integer to the DimIdxPat record, representing its original dimension.

Reply:
Maybe the function could be `[DimIdxPat rep] -> [Int]`.

Reply:
Fixed in [c29d5b](https://github.com/diku-dk/futhark/pull/2027/commits/c29d5b48215886121052bdfc3bee150709f4794c)